File: src/main/java/app/application/controller/main_window/VideoPanelController.java
package app.application.controller.main_window;

import app.application.data.entities.YoutubeVideo;
import app.application.exception.CantAbortDownloadException;
import app.application.exception.ExecutorTerminationException;
import app.application.exception.InvalidVideoUrlException;
import app.application.listener.YoutubeVideoDownloadListener;
import app.application.utils.DialogManager;
import app.application.utils.DownloadExecutorHandler;
import app.application.utils.GlobalValues;
import app.application.utils.QualityLabelExtractor;
import app.application.utils.YoutubeIdExtractor;
import app.application.utils.YoutubeUrlValidator;
import app.application.utils.service.data.YoutubeVideoDataService;
import app.application.utils.service.download.YoutubeVideoDownloadService;
import javafx.beans.binding.Bindings;
import javafx.fxml.FXML;
import javafx.scene.control.Button;
import javafx.scene.control.CheckBox;
import javafx.scene.control.ComboBox;
import javafx.scene.control.Label;
import javafx.scene.control.ProgressBar;
import javafx.scene.control.TextArea;
import javafx.scene.control.TextField;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;
import javafx.scene.layout.AnchorPane;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

@Component
public class VideoPanelController {

	@FXML
	private TextField txtDownloadLink;

	@FXML
	private ImageView imgThumbnail;

	@FXML
	private AnchorPane videoPane;

	@FXML
	private Label lblVideoTitle;

	@FXML
	private ComboBox<String> boxQuality;

	@FXML
	private TextArea txtDescription;

	@FXML
	private ProgressBar downloadProgress;

	@FXML
	private CheckBox chkAudioOnly;

	@FXML
	private Button btnSearch;

	private final YoutubeVideoDataService youtubeVideoDataService;

	private final YoutubeVideoDownloadService youtubeVideoDownloadService;

	private final YoutubeIdExtractor youtubeIdExtractor;

	private final YoutubeUrlValidator youtubeUrlValidator;

	private final DialogManager dialogManager;

	private final QualityLabelExtractor qualityLabelExtractor;

	private DownloadExecutorHandler downloadExecutorHandler;

	public VideoPanelController(
					YoutubeVideoDownloadService youtubeVideoDownloadService,
					YoutubeVideoDataService youtubeVideoDataService,
					YoutubeIdExtractor youtubeIdExtractor,
					YoutubeUrlValidator youtubeUrlValidator,
					DialogManager dialogManager,
					QualityLabelExtractor qualityLabelExtractor) {
		this.youtubeVideoDownloadService = youtubeVideoDownloadService;
		this.youtubeVideoDataService = youtubeVideoDataService;
		this.youtubeIdExtractor = youtubeIdExtractor;
		this.youtubeUrlValidator = youtubeUrlValidator;
		this.dialogManager = dialogManager;
		this.qualityLabelExtractor = qualityLabelExtractor;
	}

	private YoutubeVideo tmpYoutubeVideo;

	@FXML
	private void initialize(){
		youtubeVideoDownloadService.setYoutubeDownloadListener(new YoutubeVideoDownloadListener(downloadProgress, dialogManager));
		btnSearch.disableProperty().bind(Bindings.isEmpty(txtDownloadLink.textProperty()));
		downloadExecutorHandler = new DownloadExecutorHandler();
	}


	public void btnSearchClick() throws InvalidVideoUrlException {
		youtubeUrlValidator.checkVideoUrl(txtDownloadLink.getText());
		tmpYoutubeVideo = youtubeVideoDataService.getYoutubeVideo(youtubeIdExtractor.getVideoIdFromLink(txtDownloadLink.getText()));
		imgThumbnail.setImage(new Image(tmpYoutubeVideo.getVideoThumbnailUrl()));
		lblVideoTitle.setText(tmpYoutubeVideo.getVideoTitle());
		refreshQualityBox(qualityLabelExtractor.getQualityLabels(tmpYoutubeVideo));
		txtDescription.setText(tmpYoutubeVideo.getVideoDescription());
		videoPane.setVisible(true);
	}

	public void btnDownloadVideoClick(){
		downloadExecutorHandler.executeTask(() -> {
			if(chkAudioOnly.isSelected()){
				youtubeVideoDownloadService.downloadAudioOnlyAsync(tmpYoutubeVideo);
			}
			else{
				youtubeVideoDownloadService.downloadVideoAsync(tmpYoutubeVideo, boxQuality.getSelectionModel().getSelectedItem());
			}
		});
	}

	public void btnAbortClick() throws CantAbortDownloadException {
		downloadExecutorHandler.killTask();
	}


	private void refreshQualityBox(List<String> listWithOptions){
		boxQuality.getItems().clear();
		boxQuality.getItems().addAll(listWithOptions);
		boxQuality.getSelectionModel().select(0);
	}


}

File: src/main/kotlin/app/application/controller/main_window/VideoPanelController.kt
package app.application.controller.main_window

import app.application.data.entities.*
import app.application.exception.CantAbortDownloadException
import app.application.exception.InvalidVideoUrlException
import app.application.listener.YoutubeVideoDownloadListener
import app.application.spring.service.*
import app.application.spring.service.data.YoutubeVideoDataService
import app.application.spring.service.download.YoutubeVideoDownloadService
import app.application.utils.*
import app.application.utils.DownloadExecutorHandler.DownloaderTask
import com.github.kiulian.downloader.model.videos.formats.VideoWithAudioFormat
import javafx.beans.binding.Bindings
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleObjectProperty
import javafx.fxml.FXML
import javafx.scene.control.*
import javafx.scene.image.ImageView
import javafx.scene.input.KeyCode
import javafx.scene.input.KeyEvent
import javafx.scene.layout.AnchorPane
import javafx.util.Callback
import org.springframework.stereotype.Component

@Component
class VideoPanelController(
    private val youtubeVideoDownloadService: YoutubeVideoDownloadService,
    private val youtubeVideoDataService: YoutubeVideoDataService,
    private val youtubeIdExtractor: YoutubeIdExtractor,
    private val youtubeUrlValidator: YoutubeUrlValidator,
    private val dialogManager: DialogManager,
    private val globalObjectHandler: GlobalObjectHandler
) {

    @FXML
    private lateinit var txtDownloadLink: TextField

    @FXML
    private lateinit var imgThumbnail: ImageView

    @FXML
    private lateinit var videoPane: AnchorPane

    @FXML
    private lateinit var lblVideoTitle: Label

    @FXML
    private lateinit var boxQuality: ComboBox<VideoWithAudioFormat>

    @FXML
    private lateinit var txtDescription: TextArea

    @FXML
    private lateinit var downloadProgress: ProgressBar

    @FXML
    private lateinit var chkAudioOnly: CheckBox

    @FXML
    private lateinit var btnSearch: Button

    @FXML
    private lateinit var lblPercent: Label

    private val downloadExecutorHandler: DownloadExecutorHandler = DownloadExecutorHandler()

    private var actualYoutubeSettingsEntity: YoutubeVideoSettingsEntity? = null

    @FXML
    private fun initialize() {
        youtubeVideoDownloadService.youtubeDownloadListener = YoutubeVideoDownloadListener(downloadProgress, lblPercent, dialogManager, globalObjectHandler)
        btnSearch.disableProperty().bind(Bindings.isEmpty(txtDownloadLink.textProperty()))
        setupQualityBoxRendering()
    }

    private fun setupQualityBoxRendering() {
        boxQuality.cellFactory = Callback<ListView<VideoWithAudioFormat>, ListCell<VideoWithAudioFormat>> {
            QualityLabelListCell()
        }
        boxQuality.buttonCell = QualityLabelListCell()
    }

    @Throws(InvalidVideoUrlException::class)
    fun btnSearchClick() {
        val url = txtDownloadLink.text.trim()
        youtubeUrlValidator.checkVideoUrl(url)
        actualYoutubeSettingsEntity = YoutubeVideoSettingsEntity(youtubeVideoDataService.getYoutubeVideo(youtubeIdExtractor.getVideoIdFromLink(url)))
        updateGui()
        videoPane.isVisible = true
    }

    fun btnDownloadVideoClick() {
        downloadExecutorHandler.executeTask(object : DownloaderTask {
            override fun execute() {
                youtubeVideoDownloadService.downloadVideoAsync(actualYoutubeSettingsEntity!!)
            }
        })
    }

    @Throws(CantAbortDownloadException::class)
    fun btnAbortClick() {
        downloadExecutorHandler.killTask()
    }

    fun searchKeyPressed(keyEvent: KeyEvent){
        if(keyEvent.code == KeyCode.ENTER && !btnSearch.isDisabled){
            btnSearchClick()
        }
    }

    private fun updateGui() {
        imgThumbnail.image = actualYoutubeSettingsEntity?.youtubeEntity?.thumbnailImage
        lblVideoTitle.text = actualYoutubeSettingsEntity?.youtubeEntity?.videoTitle
        refreshQualityBox(actualYoutubeSettingsEntity?.youtubeEntity?.videoWithAudioFormat)
        txtDescription.text = actualYoutubeSettingsEntity?.youtubeEntity?.videoDescription
        actualYoutubeSettingsEntity?.settingsEntity = SettingsEntity(SimpleObjectProperty(actualYoutubeSettingsEntity?.youtubeEntity?.videoWithAudioFormat?.get(0)), SimpleBooleanProperty(chkAudioOnly.isSelected))
        actualYoutubeSettingsEntity?.settingsEntity?.qualityProperty?.bind(boxQuality.valueProperty())
        actualYoutubeSettingsEntity?.settingsEntity?.audioOnlyProperty?.bind(chkAudioOnly.selectedProperty())
    }

    private fun refreshQualityBox(listWithOptions: List<VideoWithAudioFormat>?) {
        boxQuality.items.clear()
        listWithOptions?.let { boxQuality.items.addAll(it) }
        boxQuality.selectionModel.select(0)
    }
}